crates/sui-core/src/event_handler.rs

```diff
@@ -37,29 +45,41 @@ impl EventHandler {
     pub async fn process_events(&self, effects: &TransactionEffects, timestamp_ms: u64) {
         // serially dispatch event processing to honor events' orders.
         for event in &effects.events {
-            if let Err(e) = self.process_event(event, timestamp_ms).await {
+            if let Err(e) = self
+                .process_event(event, timestamp_ms, effects.transaction_digest)
+                .await
+            {
                 error!(error =? e, "Failed to send EventEnvelope to dispatch");
             }
         }
     }
 
-    pub async fn process_event(&self, event: &Event, timestamp_ms: u64) -> SuiResult {
+    pub async fn process_event(
+        &self,
+        event: &Event,
+        timestamp_ms: u64,
+        digest: TransactionDigest,
+    ) -> SuiResult {
         let json_value = match event {
-            Event::MoveEvent(event_obj) => {
+            Event::MoveEvent {
+                type_, contents, ..
+            } => {
                 debug!(event =? event, "Process MoveEvent.");
-                let move_struct = event_obj.to_move_struct_with_resolver(
-                    ObjectFormatOptions::default(),
-                    &self.module_cache,
-                )?;
-                Some(serde_json::to_value(&move_struct).map_err(|e| {
+                // Piggyback on MoveObject's conversion logic.
+                let move_object = MoveObject::new(type_.clone(), contents.clone());
+                let layout =
+                    move_object.get_layout(ObjectFormatOptions::default(), &self.module_cache)?;
```

Comment:
Couldn't we put all this code into `to_move_struct_with_resolver` or the equivalent in MoveObject?  Just wondering if we can avoid duplicating all this code